File: Data Structures & Algorithms/max-area-of-island/submission-4.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        ROW, COL = len(grid), len(grid[0])
        maxArea = 0

        def dfs(r, c):
            if min(r, c) < 0 or r == ROW or c == COL or grid[r][c] != 1:
                return 0

            count = 1
            grid[r][c] = 0

            for dr, dc in dirs:
                count += dfs(r + dr, c + dc)

            return count

        for r in range(ROW):
            for c in range(COL):
                if grid[r][c] == 1:
                    maxArea = max(maxArea, dfs(r, c))

        return maxArea

File: Data Structures & Algorithms/max-water-container/submission-4.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int currMax = 0;
        int l = 0;
        int r = heights.size() - 1;

        while (l < r) {
            int distance = r - l;
            int minHeight = min(heights[l], heights[r]);
            currMax = max(currMax, minHeight * distance);

            if (heights[l] < heights[r]) {
                l++;
                continue;
            } else if (heights[l] > heights[r]) {
                r--;
                continue;
            } else {
                r--;
            }
        } // End While

        return currMax;
    }
};
